File: java/src/hashing/AssignmentFindPattern.java
package hashing;

import java.util.Scanner;
import java.util.List;
import java.util.ArrayList;

public class AssignmentFindPattern {
    private int p = 1000000007;

    private int x = 31;

    private long polyHash(String str) {
        long hash = 0;

        for (int i = str.length() - 1; i >= 0; i--)
            hash = (hash * this.x + ((int) str.charAt(i))) % this.p;

        return hash;
    }

    private Long[] preprocessHashes(String T, int pLen) {
        int idx = T.length() - pLen;
        Long[] H = new Long[idx + 1];
        H[idx] = this.polyHash(T.substring(idx, idx + pLen));

        long y = 1;
        for(int i = 0 ; i < pLen ; i++)
            y = (y * this.x) % this.p;

        for(int i = idx - 1; i >= 0; i--)
            H[i] = (this.x * H[i + 1] + ((int) T.charAt(i)) - y * ((int) T.charAt(i + pLen))) % p;

        return H;
    }

    public List<Integer> naive(String T, String P) {
        List<Integer> result = new ArrayList<Integer>();
        for (int i = T.length() - P.length() ; i >= 0 ; i--)
            if ( P.equals(T.substring(i, i + P.length())) )
                result.add(0, i);
        return result;
    }

    public List<Integer> sRabinKarp(String T, String P) {
        List<Integer> result = new ArrayList<Integer>();
        long pHash = this.polyHash(P);

        for (int i = T.length() - P.length() ; i >= 0 ; i--) {
            String t = T.substring(i, i + P.length());
            long tHash = this.polyHash(t);
            if (pHash == tHash && P.equals(t))
                result.add(0, i);
        }

        return result;
    }

    public List<Integer> fRabinKarp(String T, String P) {
        List<Integer> result = new ArrayList<Integer>();
        long pHash = this.polyHash(P);
        Long[] H = this.preprocessHashes(T, P.length());

        for (int i = T.length() - P.length(); i >= 0; i--) {
            if ( pHash == H[i] && P.equals(T.substring(i, i + P.length())) )
                result.add(0, i);
        }

        return result;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        String P = scanner.next();
        String T = scanner.next();

        AssignmentFindPattern fp = new AssignmentFindPattern();
        //printList("naive", fp.naive(T, P));
        //printList("slow rabin-karp", fp.sRabinKarp(T, P));
        printList("fast rabin-karp", fp.fRabinKarp(T, P));
    }

    public static void printList(String type, List<Integer> list) {
        System.out.print(type + ": ");
        for (Integer i : list)
            System.out.print(i + " ");
        System.out.println();
    }
}

File: python/hashing/assignmentFindPattern.py
from random import *
import time

def naive(T, P):
    result = []
    for i in range(len(T) - len(P), -1, -1):
        t = T[i : i+len(P)]
        if P == t:
            result.insert(0, i)
    return result

def polyHash(S, p, x):
    h = 0
    for i in range(len(S) - 1, -1, -1):
        h = (h * x + ord(S[i])) % p
    return h

def sRabinKarp(T, P, p, x):
    result = []
    pHash = polyHash(P, p, x)

    for i in range(len(T) - len(P), -1, -1):
        t = T[i : i+len(P)]
        tHash = polyHash(t, p, x)
        if pHash == tHash:
            if P == t:
                result.insert(0, i)
    return result

def preprocessHashes(T, lenP, p, x):
    idx = len(T) - lenP
    H = [None for i in range(idx + 1)]
    H[idx] = polyHash(T[idx : idx+lenP], p, x)

    y = 1
    for i in range(lenP):
        y = y * x % p

    for i in range(idx - 1, -1, -1):
        H[i] = (x * H[i + 1] + ord(T[i]) - y * ord(T[i + lenP]) ) % p
    
    return H

def fRabinKarp(T, P, p, x):
    result = []
    pHash = polyHash(P, p, x)
    H = preprocessHashes(T, len(P), p, x)

    for i in range(len(T) - len(P), -1, -1):
        if pHash == H[i]:
            if P == T[i : i+len(P)]:
                result.insert(0, i)
    return result

p = 1000000007
x = 31

'''P = input()
T = input()
#print(*naive(T, P))
#print(*sRabinKarp(T, P, p, x))
print(*fRabinKarp(T, P, p, x))'''

start_time = time.time()
print('prepare P and T: ', start_time)
P = ''
for i in range(5 * (10 ** 3)):
    P += chr(choice([randint(65, 90), randint(97, 122)]))
T = ''
for i in range(10 ** 3):
    T += P
print('start rabin karp: ', time.time() - start_time)
print(*fRabinKarp(T, P, p, x))
print('end rabin karp: ', time.time() - start_time)
